ourhexenv.py

```diff
@@ -208,14 +215,41 @@ def step(self, action):
                 self.rewards = {agent: move_r for agent in self.agents}
                 self.terminations = {agent: False for agent in self.agents}
 
-        if self.agent_selection == "player_2":
-            # Player 2 has made their first move, make pie rule unusable
-            self.is_pie_rule_usable = False
-
+        # Update cumulative rewards
         for agent in self.agents:
             self._cumulative_rewards[agent] += self.rewards[agent]
 
-        #Advance to the next agent if the game is not over
+        # Prepare observation space
+        if self.pie_rule_used:
+            # Flipping the observation space if the pie rule was used
+            flipped_board = np.where(self.board == 1, 2, np.where(self.board == 2, 1, 0))
+            self.infos = {
+                "player_1": {
+                    "observation": self.board,
+                    "pie_rule_used": 0,
+                },
+                "player_2": {
+                    "observation": flipped_board,
+                    "pie_rule_used": 0,
+                },
+            }
+        else:
+            self.infos = {
+                "player_1": {
+                    "observation": self.board,
+                    "pie_rule_used": 1 if self.is_pie_rule_usable else 0,
+                },
+                "player_2": {
+                    "observation": self.board,
+                    "pie_rule_used": 1 if self.is_pie_rule_usable else 0,
+                },
+            }
```

Comment:
We have a method called generate_info, please modify that method instead. Also, please note that each info object only passes two keywords: action_mask, and orientation. Your patch is passing "observation", and "pie_rule_used" which must be passed in the observation object instead; see the observe method where we do that.

I also see a "flipped_board" argument. I don't think this is correct. Both agents get the same board. The pie rule just "swaps" the stones from the agents. The stone previously "1" became "2", but the location and orientation of the board remain the same.

Reply:
During handling the pie rule we already performed the swap. I don't see the need to perform an extra "flip" when we already did it. Also, we are already handling returning these values in `observe` and `generate_info`

Reply:
Do you know if the pie rule is applied, do the goal edges (top to bottom or left to right) also change for the players? To my best knowledge they do, and currently it seems like we're only swapping pieces here and not the observation space. So should we transpose the board before swapping is pie rule is being used? Does this make sense?

Reply:
I think this animation provided by @SaiPrakash2k helps a lot: https://www.lutanho.net/play/hex.html. In handling the pie rule, you see that we detect the row, col where the current red token is. We then do (row, col) = 0, (col, row) = 2. This basically "transposes the stone in place". But, if you look at the animation, the board is never "transposed" as an observation space. The location of the red and blue edges remains constant throughout the game. So, I don't think we have to transpose the board, we just have to transpose the stone  - which we are already doing

Reply:
Sorry I'm late to the discussion, but yes thanks @SaiPrakash2k for the excellent resource. Let's use this as our ground truth interpretation of the pie rule. 

I had mentioned [the two distinct interpretations](https://github.com/sjsu-interconnect/ourhexgame/pull/22#discussion_r1847110898) of the pie rule in an earlier PR, so @ChristoferBerruz 
is definitely warranted in expressing his perspective. My choice of terminology may conflict with the descriptions the two of you have provided, but the presented ideas are much the same.

Let's stick this approach.

Reply:
seems to me that we only need to update the `infos` variable append condition. And Pie rule is being handled already by mirroring the positions, correct?